### ISSUE TITLE
Prediction should save intermediate results

### ISSUE THREAD

**Issue by sammlapp** [kitzeslab/opensoundscape#272]
Prediction on large datasets can take hundreds of cpu-hours, so results should be saved periodically to avoid losing all progress if there's an exception. For instance, there could be an option to append results to a file after each batch finishes.

**Comment by sammlapp** [comment]
note that as of 4/22/2021, the prediction method uses a SafeDataset meaning that bad samples will be substituted with a _different_ sample in the dataloader - at the end of prediction the scores produced by these bad files (which are actually scores for totally different samples) are replaced with NaNs. If we save intermediate results, we'll need to watch out for this.

**Comment by sammlapp** [comment]
A provisional solution to this is to split prediction into several jobs, each of which saves results to .csv when it completes. The output csv's can easily be concatenated later. 

A full solution will probably require updating the SafeDataset approach

**Comment by sammlapp** [comment]
I think an intermediate solution between separate jobs and integration within model.predict() will be good:
- slice indices
- create a dataset for each slice, predict, save
- aggregate results

Something like this:
```
from torch.utils.data import Subset
n_splits = 400
split_index = np.array_split(np.arange(len(prediction_df)),n_splits)
save_path = '.'

ds = AudioToSpectrogramPreprocessor(prediction_df.head(1),out_shape=img_shape)
ds.actions.bandpass.set(min_f=2000,max_f=8500)
ds.actions.to_spec.set(window_samples=1024,overlap_samples=512)
ds.actions.trim_audio.set(extend=True,audio_length=5.00)

for split_i, subset_idx in enumerate(split_index):
    ds = ds.sample(n=1)
    ds.df = prediction_df.iloc[subset_idx]
    split_scores, _, _ = model.predict(ds, activation_layer = None, num_workers=0)
    split_scores.to_csv(f'{save_path}/scores_split{split_i:04d}.csv')
 ```

later, concatenate csv files in bash:
```
head -1 scores_split0000.csv > full.csv
ls *.csv | xargs -n 1 tail -n +2 >> full.csv
```

**Comment by sammlapp** [comment]
Since split_and_predict can generate thousands of prediction scores on a single file, I think saving results of each file within split_and_predict might be warranted. Could look like
```
model.split_and_predict(... , save_individual_file_scores_to='./individual_file_preds_dir')
```
which would result in a csv file of predictions for each long audio file. 

Although, this might mean we should save both scores and preds (labels currently not supported)

**Comment by sammlapp** [comment]
split_and_predict is deprecated and we can continue manual looping-and-saving best practice as described above
